### PR TITLE
BigQuery Connector 

### PR DESCRIPTION
**Pull request by davidrabinowitz**
Implementing a connector for [BigQuery](https://cloud.google.com/bigquery/).

Instructions on how to use it are specified in the documentation - https://github.com/davidrabinowitz/prestosql/blob/master/presto-docs/src/main/sphinx/connector/bigquery.rst

### CODE REVIEW

File: presto-bigquery/src/test/java/io/prestosql/plugin/bigquery/MockResponsesBatch.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    @Override
+    public ReadRowsResponse next()
+    {
+        Object next = responses.poll();
+        if (next instanceof ReadRowsResponse) {
+            return (ReadRowsResponse) next;
+        }
+        else if (next instanceof RuntimeException) {
```
Comment (ebyhr) [inline presto-bigquery/src/test/java/io/prestosql/plugin/bigquery/MockResponsesBatch.java]: The same as the above.

File: pom.xml [hunk 2]
```diff
@@ -990,6 +997,104 @@
                 <version>8.0.12-2</version>
             </dependency>
 
+            <dependency>
+                <groupId>com.google.api</groupId>
```
Comment (ebyhr) [inline pom.xml]: Could you move these dependencies to `presto-bigquery/pom.xml`?

**Comment by davidrabinowitz** [comment]
After making the POM changes I have this warning I can't rid off:
```
[INFO] Found Resolved Dependency/DependencyManagement mismatches:
[INFO] 	Ignoring Direct Dependencies.
[INFO] com.fasterxml.jackson.core:jackson-core:jar was excluded in DepMgt, but version 2.10.0 has been found in the dependency tree.
[INFO] org.apache.httpcomponents:httpclient:jar was excluded in DepMgt, but version 4.5.10 has been found in the dependency tree.
[INFO] org.apache.httpcomponents:httpcore:jar was excluded in DepMgt, but version 4.4.12 has been found in the dependency tree.
[INFO] org.slf4j:slf4j-api:jar was excluded in DepMgt, but version 1.7.29 has been found in the dependency tree.
[INFO] joda-time:joda-time:jar was excluded in DepMgt, but version 2.10.5 has been found in the dependency tree.
[WARNING] Potential problems found in Dependency Management 
```

Any idea?

**Comment by electrum** [comment]
Don't worry about those dependency management warnings. They're bogus and can be ignored. I have a PR to remove that checker.

File: presto-bigquery/pom.xml [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+        <dependency>
+            <groupId>io.prestosql</groupId>
+            <artifactId>presto-tests</artifactId>
+            <scope>test</scope>
+        </dependency>
+
+        <!-- RequireUpperBoundDeps -->
+
+        <dependency>
```
Comment (electrum) [inline presto-bigquery/pom.xml]: You should be able to remove this, since it's not used. Don't worry about the bogus dependency management warning.

File: presto-bigquery/pom.xml [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+            <groupId>com.google.cloud</groupId>
+            <artifactId>google-cloud-core-http</artifactId>
+            <version>${dep.google-cloud-core.version}</version>
+        </dependency>
+
+        <dependency>
+            <groupId>com.google.code.findbugs</groupId>
+            <artifactId>jsr305</artifactId>
+            <version>3.0.2</version>
```
Comment (electrum) [inline presto-bigquery/pom.xml]: Remove this version (it's specified in the base POM) and make it `<optional>true</optional>` like we do elsewhere.

File: presto-docs/src/main/sphinx/connector/bigquery.rst [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+By default the connector creates one partition per 400MB in the table being read (before filtering). This should roughly correspond to the maximum number of readers supported by
+the BigQuery Storage API. This can be configured explicitly with the ``parallelism`` property. BigQuery may limit the number of partitions based on server constraints.
+
+Reading From Views
+^^^^^^^^^^^^^^^^^^
+The connector has a preliminary support for reading from `BigQuery views <https://cloud.google.com/bigquery/docs/views-intro>`_. Please note there are a few caveats:
+
+* BigQuery views are not materialized by default, which means that the connector needs to materialize them before it can read them. This process affects the read performance.
+* The materialization process can also incur additional costs to your BigQuery bill.
+* By default, the materialized views are created in the same project and dataset. Those can be configured by the optional ``view-materialization-project``
```
Comment (electrum) [inline presto-docs/src/main/sphinx/connector/bigquery.rst]: Prefix all the config names with `bigquery.` to match the actual property name (so users can copy/paste)

File: presto-main/etc/config.properties [hunk 6]
```diff
@@ -46,7 +46,8 @@ plugin.bundles=\
   ../presto-postgresql/pom.xml, \
   ../presto-thrift/pom.xml, \
   ../presto-tpcds/pom.xml, \
-  ../presto-google-sheets/pom.xml
+  ../presto-google-sheets/pom.xml, \
+  ../presto-bigquery/pom.xml
```
Comment (electrum) [inline presto-main/etc/config.properties]: Move this above `google-sheets` in the list so that the diff is cleaner

File: presto-bigquery/pom.xml [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+    <description>Presto - BigQuery Connector</description>
+    <packaging>presto-plugin</packaging>
+
+    <properties>
+        <air.main.basedir>${project.parent.basedir}</air.main.basedir>
+        <dep.gax.version>1.49.1</dep.gax.version>
+        <dep.grpc.version>1.24.1</dep.grpc.version>
+        <dep.google-auth-library.version>0.18.0</dep.google-auth-library.version>
+        <dep.google-cloud-core.version>1.91.3</dep.google-cloud-core.version>
+        <dep.bigquery.version>1.101.0</dep.bigquery.version>
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-bigquery/pom.xml]: I left the properties that multiple related dependencies use the same version

File: presto-bigquery/pom.xml [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+        <dep.bigquerystorage.version>0.119.0-beta</dep.bigquerystorage.version>
+    </properties>
+
+    <dependencyManagement>
+        <dependencies>
+            <dependency>
+                <groupId>com.google.api.grpc</groupId>
+                <artifactId>proto-google-common-protos</artifactId>
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-bigquery/pom.xml]: Without it the build fails with the following error:
```
[WARNING] Rule 0: org.apache.maven.plugins.enforcer.RequireUpperBoundDeps failed with message:
Failed while enforcing RequireUpperBoundDeps. The error(s) are [
Require upper bound dependencies error for com.google.api.grpc:proto-google-common-protos:1.15.0 paths to dependency are:
+-io.prestosql:presto-bigquery:330-SNAPSHOT
  +-com.google.api:gax-grpc:1.49.1
    +-com.google.api.grpc:proto-google-common-protos:1.15.0
and
+-io.prestosql:presto-bigquery:330-SNAPSHOT
  +-com.google.cloud:google-cloud-core:1.91.3
    +-com.google.api.grpc:proto-google-common-protos:1.17.0
and
+-io.prestosql:presto-bigquery:330-SNAPSHOT
  +-com.google.api.grpc:proto-google-cloud-bigquerystorage-v1beta1:0.84.0
    +-com.google.api.grpc:proto-google-common-protos:1.17.0
and
+-io.prestosql:presto-bigquery:330-SNAPSHOT
  +-com.google.api:gax-grpc:1.49.1
    +-io.grpc:grpc-protobuf:1.24.1 (managed) <-- io.grpc:grpc-protobuf:1.23.0
      +-com.google.api.grpc:proto-google-common-protos:1.12.0
and
+-io.prestosql:presto-bigquery:330-SNAPSHOT
  +-com.google.cloud:google-cloud-core:1.91.3
    +-com.google.api.grpc:proto-google-iam-v1:0.13.0
      +-com.google.api.grpc:proto-google-common-protos:1.16.0
```

File: presto-docs/src/main/sphinx/connector/bigquery.rst [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+To configure the BigQuery connector, create a catalog properties file
+in ``etc/catalog`` named, for example, ``bigquery.properties``, to
+mount the BigQuery connector as the ``bigquery`` catalog.
+Create the file with the following contents, replacing the
+connection properties as appropriate for your setup:
+
+.. code-block:: none
+
+    connector.name=bigquery
+    project-id=<your Google Cloud Platform project id>
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-docs/src/main/sphinx/connector/bigquery.rst]: Yes if you refer to projects other than the billed project (or the parent project). It is very common among GCP users to have different GCP projects per group or per product, in order to have better control over budget and for better governance. However, in many cases, users of one project may get access to resources of another project - think that marketing campaigns may get access to some tables of the analytics group for example.

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryColumnHandle.java [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    @JsonProperty
+    public BigQueryType getBigQueryType()
+    {
+        return bigQueryType;
+    }
+
+    @Override
+    public ImmutableMap<String, BigQueryType.Adaptor> getBigQuerySubTypes()
+    {
+        return subColumns.stream().collect(ImmutableMap.toImmutableMap(BigQueryColumnHandle::getName, column -> column));
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryColumnHandle.java]: Thanks for the tip, but because of typing issues uniqueIndex Fails.

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryPageSource.java [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+                appendTo(type.getTypeParameters().get(index), record.get(fieldNames.get(index)), builder);
+            }
+            output.closeEntry();
+            return;
+        }
+        else {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Unhandled type for Block: " + type.getTypeSignature());
+        }
+
+        // not a convertible value
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryPageSource.java]: Also the throw ? BTW, the code was adapted from MongoPageSource.java

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryColumnHandle.java [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+
+    @Override
+    @JsonProperty
+    public BigQueryType getBigQueryType()
+    {
+        return bigQueryType;
+    }
+
+    @Override
+    public ImmutableMap<String, BigQueryType.Adaptor> getBigQuerySubTypes()
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryColumnHandle.java]: Hi @electrum , I agree that returning a base class rather than the implementation is a good idea, but in this case I'm not sure it applies. `ImmutableMap` is not really a `Map` implementation, as it lacks a lot of its functionality. If we go by the [LSP](https://en.wikipedia.org/wiki/Liskov_substitution_principle) then `Map` is a subclass of `ImmutableMap` as it provides all of `ImmutableMap` functionality, but the opposite is not true and can lead to an `UnsupportedOperationException`. What do you think?

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryColumnHandle.java [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+    @JsonCreator
+    public BigQueryColumnHandle(
+            @JsonProperty("name") String name,
+            @JsonProperty("bigQueryType") BigQueryType bigQueryType,
+            @JsonProperty("mode") Field.Mode mode,
+            @JsonProperty("subColumns") List<BigQueryColumnHandle> subColumns,
+            @JsonProperty("description") String description)
+    {
+        this.name = requireNonNull(name, "column name cannot be null");
+        this.bigQueryType = requireNonNull(bigQueryType, "column type cannot be null for column [" + name + "]");
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryColumnHandle.java]: @electrum It fails with `Prefer java.util.Objects.requireNonNull(Object, String) or java.util.Objects.requireNonNull(Object, Supplier<String>)`

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryConfig.java [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+        Optional<Credentials> credentialsFromKey = credentialsKey.map(this::createCredentialsFromKey);
+        Optional<Credentials> credentialsFromFile = credentialsFile.map(this::createCredentialsFromFile);
+        return Stream.of(credentialsFromKey, credentialsFromFile)
+                .flatMap(Streams::stream)
+                .findFirst();
+    });
+
+    public Optional<Credentials> createCredentials()
+    {
+        if (credentialsKey.isPresent() && credentialsFile.isPresent()) {
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryConfig.java]: @electrum The actual check (which I've just noticed) is that the Credentials can be created from the environment - either by setting the GOOGLE_APPLICATION_CREDENTIALS env variable or by running on Google Compute Engine with the right role. However such test I'm afraid will fail on the CI/CD

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryColumnHandle.java [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+        this.mode = mode != null ? mode : Field.Mode.NULLABLE;
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryColumnHandle.java]: My mistake, this field cannot be null

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryUtil.java [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+        if (t instanceof StatusRuntimeException) {
+            StatusRuntimeException statusRuntimeException = (StatusRuntimeException) t;
+            return statusRuntimeException.getStatus().getCode() == Status.Code.INTERNAL &&
+                    INTERNAL_ERROR_MESSAGES.stream()
+                            .anyMatch(errorMsg -> statusRuntimeException.getMessage().contains(errorMsg));
+        }
+        return false;
+    }
+
+    static void convertAndThrow(BigQueryError error)
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryUtil.java]: It may be used in other places

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/ReadRowsHelper.java [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+    public ReadRowsHelper(BigQueryStorageClient client, ReadRowsRequest.Builder request, int maxReadRowsRetries)
+    {
+        this.client = client;
+        this.request = request;
+        this.maxReadRowsRetries = maxReadRowsRetries;
+    }
+
+    public Iterator<ReadRowsResponse> readRows()
+    {
+        StreamPosition.Builder readPosition = request.getReadPositionBuilder();
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/ReadRowsHelper.java]: @electrum I'm sorry, but I don't understand what to inline. the readPosition is a builder used to set the correct read position in case of a retry.

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/ReadSessionCreator.java [hunk 18]
```diff
@@ -0,0 +1,232 @@
...
+    public Storage.ReadSession create(TableId table, ImmutableList<String> selectedFields, String filter, int parallelism)
+    {
+        Table tableDetails = bigquery.getTable(table);
+
+        TableInfo actualTable = getActualTable(tableDetails, selectedFields, new String[] {});
+
+        try (BigQueryStorageClient bigQueryStorageClient = bigQueryStorageClientFactory.createBigQueryStorageClient()) {
+            ReadOptions.TableReadOptions readOptions = ReadOptions.TableReadOptions.newBuilder()
+                    .addAllSelectedFields(selectedFields)
+                    //.setRowRestriction(filter)
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/ReadSessionCreator.java]: Because of the underlining NPE, I've changed the parameter to an empty String, like in the spark connector

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/ReadSessionCreator.java [hunk 19]
```diff
@@ -0,0 +1,232 @@
...
+        else {
+            // not regular table or a view
+            throw new IllegalArgumentException(format("Table type '%s' of table '%s.%s' is not supported",
+                    tableType, table.getTableId().getDataset(), table.getTableId().getTable()));
+        }
+    }
+
+    String createSql(TableId table, ImmutableList<String> requiredColumns, String[] filters)
+    {
+        String tableName = format("%s.%s.%s", table.getProject(), table.getDataset(), table.getTable());
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/ReadSessionCreator.java]: It is not allowed in BigQuery, see [Dataset Naming](https://cloud.google.com/bigquery/docs/datasets#dataset-naming), [Table Naming](https://cloud.google.com/bigquery/docs/tables#table_naming). In general:

- Contain up to 1,024 characters
- Contain letters (upper or lower case), numbers, and underscores

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryType.java [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+{
+    BOOLEAN(BooleanType.BOOLEAN),
+    BYTES(VarbinaryType.VARBINARY),
+    DATE(DateType.DATE),
+    DATETIME(VarcharType.VARCHAR),
+    FLOAT(DoubleType.DOUBLE),
+    GEOGRAPHY(VarcharType.VARCHAR),
+    INTEGER(BigintType.BIGINT),
+    NUMERIC(DecimalType.createDecimalType(NUMERIC_DATA_TYPE_PRECISION, NUMERIC_DATA_TYPE_SCALE)),
+    RECORD(null),
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryType.java]: @electrum The BigQuery Java library use the LegacySQLType in the [Field](https://googleapis.dev/java/google-cloud-clients/latest/com/google/cloud/bigquery/Field.html) type. It is the only type that had changed name.

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryType.java [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.prestosql.plugin.bigquery.BigQueryMetadata.NUMERIC_DATA_TYPE_PRECISION;
+import static io.prestosql.plugin.bigquery.BigQueryMetadata.NUMERIC_DATA_TYPE_SCALE;
+import static io.prestosql.spi.type.VarcharType.createUnboundedVarcharType;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+public enum BigQueryType
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryType.java]: From the API stand of point, in BigQuery an Array is not a type but rather a field mode (one of nullable, required or repeated). That means there is not direct conversion from StandardSQLTypeName to an array, but you need the legacy type and the mode in order to construct to proper standard type

File: presto-bigquery/src/test/java/io/prestosql/plugin/bigquery/TestReadRowsHelper.java [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+import java.util.Arrays;
+import java.util.Iterator;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@Test
+public class TestReadRowsHelper
+{
+    // it is not used, we just need the reference
+    BigQueryStorageClient client;
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-bigquery/src/test/java/io/prestosql/plugin/bigquery/TestReadRowsHelper.java]: Replaced with a mock. We have added requiredNotNull in the constructor

File: presto-bigquery/src/test/java/io/prestosql/plugin/bigquery/TestTypeConversions.java [hunk 23]
```diff
@@ -0,0 +1,283 @@
...
+    @Test
+    public void testConvertDateField()
+    {
+        assertSimpleFieldTypeConversion(LegacySQLTypeName.DATE, DateType.DATE);
+    }
+
+    @Test
+    public void testConvertDateTimeField()
+    {
+        // TODO(davidrabinowitz): decide what type to convert it to
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-bigquery/src/test/java/io/prestosql/plugin/bigquery/TestTypeConversions.java]: I'm not so sure regarding `TIMESTAMP`. BigQuery's `DATETIME` [is not a timestamp equivalent](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#datetime-type), but rather "a `DATETIME` object does not refer to an absolute instance in time. Instead, it is the _civil time_, or the time that a user would see on a watch or calendar.

[the remark this answers is not in the document]
Comment (findepi) [reply presto-bigquery/src/test/java/io/prestosql/plugin/bigquery/TestTypeConversions.java]: @davidrabinowitz see https://github.com/prestosql/presto/issues/37 for information about what Presto TIMESTAMP is and should be.

Anyway, "a DATETIME object does not refer to an absolute instance in time. Instead, it is the civil time, or the time that a user would see on a watch or calendar." matches exactly the description of what TIMESTAMP is in eg Postgres, so you can draw inspiration from that connector. 
The bottom line is that you should have a switch on `isLegacyTimestamp` and encoded internal value appropriately. The end result should _display_ in Presto CLI the same way it does in BQ, regardless of CLI session zone and legacy timestamp semantics.

File: presto-docs/src/main/sphinx/connector/bigquery.rst [hunk 24]
```diff
@@ -0,0 +1,156 @@
...
+=========  ======================== =============================================================================================================
+
+FAQ
+---
+
+What is the Pricing for the Storage API?
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+See the `BigQuery pricing documentation <https://cloud.google.com/bigquery/pricing#storage-api>`_.
+
+Calling count(*) takes a long time
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-docs/src/main/sphinx/connector/bigquery.rst]: @electrum @findepi Perhaps I'm missing something, but currently `SELECT * FROM t` and `SELECT COUNT(*) FROM t` looks the same where I create the API query, as `applyProjection()` was not called in both cases. How do I distinct an empty row from a full row? I there some call that I have missed?

[the remark this answers is not in the document]
Comment (findepi) [reply presto-docs/src/main/sphinx/connector/bigquery.rst]: You get the list of required columns also in `io.prestosql.split.PageSourceProvider#createPageSource`.

[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-docs/src/main/sphinx/connector/bigquery.rst]: Unfortunately it is too late for me. The way the BIgQuery Storage API works is that I first open a `ReadSession`, where I need to specify the required columns. In response I get a list of streams and each split read from those streams. While I can create empty rows in the PageSource based on this, it will be only after the data has been read from the API.

Currently I create the read session in the `ConnectoeSpiltManager`, where I don't have this information. The only workaround I found so far is that I've noticed that `ConnectorMetadata.getColumnMetadata()` is called for each requested column, and use this  somehow using the session or the `ConnectorTableHandle`

[the remark this answers is not in the document]
Comment (findepi) [reply presto-docs/src/main/sphinx/connector/bigquery.rst]: > currently `SELECT * FROM t` and `SELECT COUNT(*) FROM t` looks the same where I create the API query, as `applyProjection()` was not called in both cases.

@davidrabinowitz  is this reproducible with eg a postgres connector? can you file an issue?

[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-docs/src/main/sphinx/connector/bigquery.rst]: Sorry, couldn't find it, as the postgres connector relies on the base-jdbc plugin. It's `getSplits()` relies on the default implementation, and there is no `ConnectorPageSourceProvider`

I've submitted Issue #2943 per your suggestion

File: presto-bigquery/pom.xml (hunk 8 above)
[the remark this answers is not in the document]
Comment (electrum) [reply presto-bigquery/pom.xml]: Thanks, I understand now. It looks like Maven is picking the older version based on `grpc-protobuf` for some reason.

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryPageSource.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (electrum) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryPageSource.java]: No, the throw is correct. This is just dead code (and separately should be cleaned up in the Mongo connector)

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryColumnHandle.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (electrum) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/BigQueryColumnHandle.java]: The map return here is for callers to read from. They have no reason to attempt to write to it, nor should they assume they can. Also, writing to a collection returned from a method that's not intended to be written to is bad practice (this can cause corruption or other bugs if the code was not written defensively in this regard).

We prefer to leave APIs flexible to allow changing the implementation. Using `Map` here allows the implementation to later change to `unmodifiableMap()`, an immutable map from JDK 9+, etc. This is similar to why we wouldn't declare or return `HashMap`.

You might also consider this a code style issue. I can understand the argument for using the `Immutable` interfaces everywhere possible in a code base, but we don't do that in Presto.

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/ReadRowsHelper.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (electrum) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/ReadRowsHelper.java]: I was thinking to structure the code something like this
```java
if (isRetryable) {
    request.setReadPosition(request.getReadPositionBuilder().setOffset(readRowsCount));
    serverResponses = fetchResponses(request);
    retries++;
}

[the remark this answers is not in the document]
Comment (electrum) [reply presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/ReadRowsHelper.java]: Actually, is it required to call `set`? My understanding is that the builder is tied to the parent object and will be used when building the parent.
```java
request.getReadPositionBuilder().setOffset(readRowsCount);
serverResponses = fetchResponses(request);

File: presto-docs/src/main/sphinx/connector/bigquery.rst (hunk 24 above)
[the remark this answers is not in the document]
Comment (electrum) [reply presto-docs/src/main/sphinx/connector/bigquery.rst]: I commented on the issue. `applyProjection()` seems to be called as expected with an empty projection list for this case.

[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply presto-docs/src/main/sphinx/connector/bigquery.rst]: Thanks for the tip, it is implemented now by running count(*) on BigQuery

**Comment by davidrabinowitz** [comment]
@electrum @findepi All the issues have been fixed

File: presto-bigquery/src/test/java/io/prestosql/plugin/bigquery/BigQueryQueryRunner.java [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+                .setCatalog("bigquery")
+                .setSchema(TPCH_SCHEMA)
+                .build();
+    }
+
+    public static void main(String[] args)
+            throws Exception
+    {
+        Logging.initialize();
+        DistributedQueryRunner queryRunner = createQueryRunner();
```
Comment (electrum) [inline presto-bigquery/src/test/java/io/prestosql/plugin/bigquery/BigQueryQueryRunner.java]: We should set this on port 8080 like the others do. This makes it easy to connect for testing.

File: presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/Conversions.java [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+        FieldList subFields = field.getSubFields();
+        List<BigQueryColumnHandle> subColumns = subFields == null ?
+                Collections.emptyList() :
+                subFields.stream()
+                        .map(Conversions::toColumnHandle)
+                        .collect(Collectors.toList());
+        return new BigQueryColumnHandle(
+                field.getName(),
+                BigQueryType.valueOf(field.getType().name()),
+                field.getMode(),
```
Comment (electrum) [inline presto-bigquery/src/main/java/io/prestosql/plugin/bigquery/Conversions.java]: This seems to return null rather than `NULLABLE` which causes NPE.

**Comment by electrum** [comment]
Merged, thanks!
